WarpX: Python on pyAMReX

ax3l opened the pull request:
Long overdue update for WarpX: in 2024, we updated our Python bindings to rely on the new pyAMReX package. This deprecates the old `py-warpx` package and adds a new dependency and variant to WarpX. This update also removes any vendored (auto-downloaded) packages from our superbuilds (i.e., AMReX) and properly uses external dependencies now.

Add @EZoni to WarpX package co-maintainers.

Also deprecates old versions that we will not continue to support.

- [x] needs 24.08 releases for
  - https://github.com/AMReX-Codes/pyamrex/pull/339
  - https://github.com/ECP-WarpX/WarpX/pull/5055
  - https://github.com/ECP-WarpX/impactx/pull/644

ax3l committed:
WarpX: Python on pyAMReX

Long overdue update for WarpX: in 2024, we updated our Python
bindings to rely on the new pyAMReX package. This deprecates the old
`py-warpx` package and adds a new dependency and variant to WarpX.

Also deprecates old versions that we will not continue to support.

ax3l commented:
@spack/e4s for the WarpX E4S Packages, we can remove the `py-warpx` packages now but should add the `+python` variant to the deployed packages. `+python` is non-default in Spack packages by convention and thus `default=False` in the `warpx` package, too.

eugeneswalker commented:
> @spack/e4s for the WarpX E4S Packages, we can remove the `py-warpx` packages now but should add the `+python` variant to the deployed packages.

I'm confused. We should remove py-warpx but also add +python to the deployed packages??

ax3l commented:
Correct.

ax3l commented on var/spack/repos/builtin/packages/py-amrex/package.py:
```diff
@@ -63,46 +72,46 @@ class PyAmrex(PythonPackage, CudaPackage, ROCmPackage):
     depends_on("py-cupy", type="test", when="+cuda")
 
     tests_src_dir = "tests/"
+    install_targets = ["install", "pip_install"]
```
This seems to install into the site packages of the `python-venv` install location, not the `py-amrex` install location. It works, but is that what we want? Otherwise, I can set more flags if we want this under the `py-amrex` install prefix.

This is what it calls: https://github.com/AMReX-Codes/pyamrex/blob/24.07/CMakeLists.txt#L302

ax3l replied:
@adamjstewart do you know what I need to do to a `pip install` command to ensure it's local and not in the global venv?

adamjstewart replied:
You need to specify the `--prefix`. Look at the `PythonPackage` base class for details. Also make sure to use `PythonPipBuilder.std_args` and extend the `PythonExtension` base class.

ax3l replied:
Thank you so much, Adam  -- added now :)

ax3l commented:
With the latest patches, `spack install -v warpx@develop +python` already builds.

ax3l commented on var/spack/repos/builtin/packages/py-warpx/package.py:
```diff
@@ -21,34 +23,114 @@ class PyWarpx(PythonPackage):
+    version(
+        "23.08",
+        sha256="67695ff04b83d1823ea621c19488e54ebaf268532b0e5eb4ea8ad293d7ab3ddc",
+        deprecated=True,
```
It looks like CI which runs E4S packages, which should be updated after this PR in terms of specs built, does not like this. We can keep for now the last old release if needed:
```suggestion
```

ax3l commented on var/spack/repos/builtin/packages/warpx/package.py:
```diff
@@ -3,62 +3,176 @@
+    version(
+        "23.08",
+        sha256="67695ff04b83d1823ea621c19488e54ebaf268532b0e5eb4ea8ad293d7ab3ddc",
+        deprecated=True,
```
It looks like CI which runs E4S packages, which should be updated after this PR in terms of specs built, does not like this. We can keep for now the last old release if needed:
```suggestion
```

eugeneswalker commented:
> @eugeneswalker should I update the E4S changes we discussed above in the `share/spack/gitlab/cloud_pipelines/` files of this repo with this PR? Or are the E4S recipes maintained in an external repo?

Yes, will you please update them here? We have recipes elsewhere and I can take care of those based on the changes you make to the pipelines in Spack repo.

ax3l committed:
Update Cloud/E4S Pipelines for WarpX

`py-warpx` is replaced by `warpx +python`
oneAPI does not support IPO/LTO: diable for `py-amrex` even though
pybind11 strongly encourages it.

ax3l commented:
Thanks for the guidance, pushed now.

ax3l commented:
@eugeneswalker @tgamblin do you know if the compilers in the oneAPI CI runners for E4S are properly set up (e.g., via `source /opt/intel/oneapi/setvars.sh`)? I think the _linker_ might not properly set because I observe that CMake's IPO tests fail.

In my mainline repo of pyAMReX, I [build with oneAPI using LTO/IPO for python bindings](https://github.com/AMReX-Codes/pyamrex/blob/development/.github/workflows/intel.yml), which is a practice [highly encouraged by our dependency pybind11](https://pybind11.readthedocs.io/en/stable/compiling.html#building-with-cmake) due to the way it generates code. (I do not care much about general IPO support beyond that and understand that Spack defaults to off for this.)

I disabled IPO/LTO for `py-amrex` for now on the oneAPI runners to make the tests pass, but I think this can be fixed in Spack because I don't have the same issue with the oneAPI build chain outside of Spack.

ax3l commented:
Ready to merge :tada:
